src/Microsoft.AspNetCore.Analyzer.Testing/DiagnosticVerifier.cs

```diff
@@ -0,0 +1,195 @@
+    {
+        internal static string DefaultFilePathPrefix = "Test";
+        internal static string TestProjectName = "TestProject";
+
+        /// <summary>
+        /// Given classes in the form of strings, their language, and an IDiagnosticAnalyzer to apply to it, return the diagnostics found in the string after converting it to a document.
+        /// </summary>
+        /// <param name="sources">Classes in the form of strings</param>
+        /// <param name="analyzer">The analyzer to be run on the sources</param>
+        /// <param name="additionalEnabledDiagnostics">Additional diagnostics to enable at Info level</param>
+        /// <returns>An IEnumerable of Diagnostics that surfaced in the source code, sorted by Location</returns>
+        protected Task<Diagnostic[]> GetSortedDiagnosticsAsync(string[] sources, DiagnosticAnalyzer analyzer, string[] additionalEnabledDiagnostics)
+        {
+            return GetSortedDiagnosticsFromDocumentsAsync(analyzer, GetDocuments(sources), additionalEnabledDiagnostics);
+        }
+
+        /// <summary>
+        /// Given an analyzer and a document to apply it to, run the analyzer and gather an array of diagnostics found in it.
+        /// The returned diagnostics are then ordered by location in the source document.
+        /// </summary>
+        /// <param name="analyzer">The analyzer to run on the documents</param>
+        /// <param name="documents">The Documents that the analyzer will be run on</param>
+        /// <param name="additionalEnabledDiagnostics">Additional diagnostics to enable at Info level</param>
+        /// <returns>An IEnumerable of Diagnostics that surfaced in the source code, sorted by Location</returns>
+        protected static async Task<Diagnostic[]> GetSortedDiagnosticsFromDocumentsAsync(DiagnosticAnalyzer analyzer, Document[] documents, string[] additionalEnabledDiagnostics)
+        {
+            var projects = new HashSet<Project>();
+            foreach (var document in documents)
+            {
+                projects.Add(document.Project);
+            }
+
+            var diagnostics = new List<Diagnostic>();
+            foreach (var project in projects)
+            {
+                var compilation = project.GetCompilationAsync().Result;
+
+                // Enable any additional diagnostics
+                var options = compilation.Options;
+                if (additionalEnabledDiagnostics.Length > 0)
+                {
+                    options = compilation.Options
+                        .WithSpecificDiagnosticOptions(
+                            additionalEnabledDiagnostics.ToDictionary(s => s, s => ReportDiagnostic.Info));
+                }
+
+                var compilationWithAnalyzers = compilation
+                    .WithOptions(options)
+                    .WithAnalyzers(ImmutableArray.Create(analyzer));
+
+                var diags = await compilationWithAnalyzers.GetAllDiagnosticsAsync();
```

Comment:
Isn't there a `GetAnalyzerDiagnosticsAsync()`? You could avoid the filtering stuff with it?

Reply:
You would want to see compilation errors if you have any.

Reply:
So the way I did this in Mvc was just have the source code checked as individual files. That way you just got first class completion, compiler errors etc. e.g. https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNetCore.Mvc.Analyzers.Test/TestFiles/DiagnosticsAreReturned_ForUseOfHtmlPartial.cs

It also ensures that you can verify at build time that you aren't breaking the test file due to changes in Sdk \ references etc.

Reply:
I want to generate a lot of combinations of source files (see https://github.com/aspnet/Logging/blob/dev/test/Microsoft.Extensions.Logging.Analyzer.Test/FormatStringAnalyzerTests.cs#L48). Checking them in wouldn't work in this case.

Reply:
We can leave this in, just suggesting a pattern for analyzer tests.